### ISSUE TITLE
Q: creating for sending other than to the browser

### ISSUE THREAD

**Issue by jrochkind** [WeTransfer/zip_tricks#45]
This looks pretty nifty, you figured out some fancy zip stuff. 

My use case is not sending to the browser, but sending... elsewhere. Actually to S3, ideally streamed as the zip is created, but made more complicated by the fact that the ruby AWS SDK doesn't currently have the right API for this. 

But ignoring that for now, is there any way to use the Streamer, not with an IO object, but with a block that gets yielded to with "chunks", serially, so I can send the chunks wherever I want to store the zip?

Does this make sense?

**Comment by julik** [comment]
**TL;DR** Yes, makes sense and there are facilities, but you will have some wiring to do on your end.

If you are working with S3 specifically you will need some extra machinery for that, because S3 does not support appends. You probably should use multipart uploads, which have a limit of all parts except the last one being 5MB or larger. This means that you have to do some buffering on the zip output until your part is large enough to do the multipart PUT. This is one way to go about it of course. 

For capturing the output you can use `ZipTricks::BlockWrite` which is made just for that.
Now, if you go for the multipart PUT approach your buffering code should go inside the block

```
part_n = 1
w = ZipTricks:BlockWrite.new do |binary_string|
    # binary_string is of _arbitrary_ size since we do no buffering in zip_tricks!
    # can even be 1 byte. no guarantees.
    some_buf << binary_string
    if some_buf.size >= 5*1024*1024
       multipart_upload.part(part_n).put(some_buf) # I forgot that API but you will also need a part counter
       part_n += 1
       some_buf.clear
    end
end

ZipTricks::Streamer.open(w) do |...|
end

# check if there's anything left over
if some_buf.size > 0 # last part
  multipart_upload.part(part_n).put(some_buf)
end
```

Or you could write your own wrapper, like so:

```
class MySink
  def <<(data)
     # do whatever you like
  end
end

ZipTricks::Streamer.new(MySink.new) do |...|
end
```

You can also "wrap it inside out" by having an object that can be `read` and instead will return the last string given by ZipTricks, or use a pipe

```
r, w = IO.pipe
your_http_client.put(presigned_s3_url, body: r) # This call should not block, or be in another thread, or something
ZipTricks::Streamer.open(w) do |...|
end
```

It is possible though that this contraption will deadlock in some nasty manner somewhere, _and_ you will need your HTTP client library to support chunked requests - I understand these are possible but I never used them in the wild. This also might play up with the signed Content-Length request header on S3 which I _think_ they require, but got no time to check now. For example - apparently requests for Python does support chunked requests with a deferred reader for body. I believe HTTP.rb does as well. The HTTP client library will then have to apply the chunked encoding on top of the strings you give it, since you don't know how large your request is going to be. And your request sending can be interrupted, whereupon you will have to restart the ZIP creation again.

**Comment by jrochkind** [comment]
Thanks, this is helpful. 

There is actually a not-yet-merged PR in AWS ruby sdk to allow multi-part uploads in a streaming way. https://github.com/aws/aws-sdk-ruby/pull/1711

While not yet merged (although hopefully it will be merged soon, they have expressed an intention to merge), it's mostly code you can easily copy-and-paste adapt to be just third-party code that uses the gem. 

Combining your example and that should result in something fairly reliable. I am not sure if I am going to do it at present, it's a lot of hackery for a feature that may not be worth the value to me. 

But thank you for the response, and hopefully this issue will be helpful to someone in the future that finds it!

**Comment by julik** [comment]
Come to think of it you can even (mis)use `RackBody` for this - it converts the ZipTricks writes to the output of `each`:

```
deferred_zip = ZipTricks::RackBody.new do | zip |
  zip.write_stored_file('mov.mp4') do |sink| # Those MPEG4 files do not compress that well
    File.open('mov.mp4', 'rb'){|source| IO.copy_stream(source, sink) }
  end
  zip.write_deflated_file('long-novel.txt') do |sink|
    File.open('novel.txt', 'rb'){|source| IO.copy_stream(source, sink) }
  end
end
deferred_zip.each do |binary_chunk| # this can also be fibered/suspended using an Enumerator
   ...
end
```

**Comment by julik** [comment]
@jrochkind 🎩 my pleasure! Thanks for using ZipTricks.

**Comment by jrochkind** [comment]
With the RAckBody version... what makes it "deferred", is there threading involved i guess?

Trying to think through where and when those bytes are actually produced.... I guess if it's producing them quicker than the "sink" (S3 streamed multi-part upload) can read them, you could end up with a bunch buffered in memory? (I have REALLY large zips, 10s of gigs, is why this whole thing comes up, heh).

**Comment by julik** [comment]
> With the RackBody version... what makes it "deferred", is there threading involved i guess?

No, not really. The block that you give it (where you do calls to `.write_deflated_file` etc.) will only start to execute once you call `each` on it. It captures the block given to the constructor and executes it on-demand. So no magic stuff.

**Comment by jrochkind** [comment]
But if you tell it to write like 100 stored or deflated files... it won't actually write them until the bytes from preceding writes have been consumed by the 'each'? It doesn't write everything into some memory buffer, but pauses not writing until something asks for more bytes. That's perfect if I understand right.

**Comment by julik** [comment]
@jrochkind I have added an example of how you could do this in 5bb1b065529ffbbfee87e2103de2f5fb7f717b5d

If this proves helpful, let me know so that I can close the issue 🍾

**Comment by jrochkind** [comment]
yes, thanks!
